### PR TITLE
Use same logic as `-e` when stdin is not a terminal.

### PR DESCRIPTION
**Pull request by mberhault**
This allows `echo "show databases" | ./cockroach sql` to work. Before
this, we would need a trailing `;`.

Fixes 5983.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5988)

### CODE REVIEW

File: cli/sql.go [hunk 1]
```diff
@@ -271,6 +272,15 @@ func runTerm(cmd *cobra.Command, args []string) error {
 	if len(cliContext.execStmts) > 0 {
 		// Single-line sql; run as simple as possible, without noise on stdout.
 		return runStatements(conn, cliContext.execStmts)
+	} else if !isatty.IsTerminal(os.Stdin.Fd()) {
+		// Non-tty stdin (eg: commands piped into cockroach): run quietly.
+		// TODO(marc): we may have to read through until ';' is encountered, but
```
Comment (bdarnell) [inline cli/sql.go]: What do you mean by this comment? Just reading a statement at a time instead of reading the whole thing at once?

**Comment by bdarnell** [comment]
LGTM

File: cli/sql.go [hunk 2]
```diff
@@ -271,6 +272,15 @@ func runTerm(cmd *cobra.Command, args []string) error {
 	if len(cliContext.execStmts) > 0 {
 		// Single-line sql; run as simple as possible, without noise on stdout.
 		return runStatements(conn, cliContext.execStmts)
+	} else if !isatty.IsTerminal(os.Stdin.Fd()) {
```
Comment (tschottdorf) [inline cli/sql.go]: will this cause invocations as in #5981 to lose their buffering? If so, :-1:. I don't want my 10gb data dump to be held in memory before executing.

File: cli/sql.go (hunk 1 above)
Comment (mberhault) [inline cli/sql.go]: yup. and Tobi actually pointed out an example before I thought we had one. So this is a no-go.
The proper solution is to pass a reader and parse out individual (or blocks of) statements. Unfortunately, to do this properly, we probably need to go reach into the parser.

File: cli/sql.go (hunk 2 above)
Comment (mberhault) [inline cli/sql.go]: you're right. I guess my TODO needs to be done first.

Comment (mberhault) [inline cli/sql.go]: have you encountered any issues with our line parsing? (eg: looking for a `;\n` to terminate a statement)
I can do the same thing right now, and leave proper parsing as a todo.

**Comment by mberhault** [comment]
PTAL: this now uses the same parsing as the interactive shell.
when not in interactive mode (stdin or stdout are not TTYs):
* don't display intro text
* execute last built-up command on EOF
* don't load/save history
* exit on first error

File: cli/sql.go (hunk 2 above)
Comment (tschottdorf) [inline cli/sql.go]: No, that works for me (I'm doing only single-line statements).

File: cli/sql.go [hunk 3]
```diff
@@ -197,24 +213,30 @@ func runInteractive(conn *sqlConn) (exitErr error) {
...
+				readline.SetHistoryPath("")
+			}
 		}
 
 		if exitErr = runPrettyQuery(conn, os.Stdout, makeQuery(fullStmt)); exitErr != nil {
 			fmt.Fprintln(osStderr, exitErr)
 		}
 
+		if !isInteractive && exitErr != nil {
+			break
```
Comment (tschottdorf) [inline cli/sql.go]: So this ends processing when an error is encountered? Hmm, I suppose that's safe though it does mean I need to change my example (stupid reasons). Fine by me, though `psql` definitely just keeps going so maybe so should we.

**Comment by tschottdorf** [comment]
LGTM

File: cli/sql.go (hunk 3 above)
Comment (mberhault) [inline cli/sql.go]: oh. it does? that seems dangerous, at least by default.
Anyway, I'm happy to keep going on errors.

Comment (tschottdorf) [inline cli/sql.go]: I'm sure they have some option for it, but in #5981 I'm making us of that for both our shell and Postgres' without specifying anything.

Comment (mberhault) [inline cli/sql.go]: ok. Reverting back to previous functionality. ie: keep going on errors.

**Comment by jseldess** [comment]
Docs updated with https://github.com/cockroachdb/docs/pull/270

**Pull request merged**